File: frontend/js/pages/invoices.js
var InvoicesPage = {
  currentPage: 1,
  search: '',
  statusFilter: '',
  _productsCache: [],

  async render() {
    return '<div class="loading"><div class="spinner"></div></div>';
  },

  async renderInvoices() {
    try {
      var params = new URLSearchParams({ page: this.currentPage, limit: 50 });
      if (this.search) params.set('search', this.search);
      if (this.statusFilter) params.set('status', this.statusFilter);

      var data = await API.get('/invoices?' + params.toString());
      var isAdmin = App.getUser().role === 'admin';

      var html = '<div class="page-title">Invoices</div>'
        + '<div class="action-bar">'
        + '<button class="btn btn-primary btn-sm" onclick="InvoicesPage.showForm()">+ New Invoice</button>';
      if (isAdmin) {
        html += '<button class="btn btn-outline btn-sm" onclick="InvoicesPage.exportExcel()">📤 Export</button>';
      }
      html += '</div>'
        + '<div class="search-bar">'
        + '<input type="text" placeholder="Search invoices..." id="inv-search"'
        + ' value="' + this.search + '" oninput="InvoicesPage.searchInvoices(this.value)">'
        + '<select onchange="InvoicesPage.filterStatus(this.value)">'
        + '<option value="">All</option>'
        + '<option value="Paid"' + (this.statusFilter === 'Paid' ? ' selected' : '') + '>Paid</option>'
        + '<option value="Unpaid"' + (this.statusFilter === 'Unpaid' ? ' selected' : '') + '>Unpaid</option>'
        + '<option value="Partial"' + (this.statusFilter === 'Partial' ? ' selected' : '') + '>Partial</option>'
        + '</select></div>'
        + '<div class="card"><div class="table-container"><table><thead><tr>'
        + '<th>#</th><th>Customer</th><th>Amount</th><th>Paid</th><th>Status</th><th>Actions</th>'
        + '</tr></thead><tbody>';

      if (data.invoices.length) {
        html += data.invoices.map(function(inv) {
          var statusClass = 'badge-' + (inv.payment_status || '').toLowerCase();
          return '<tr>'
            + '<td style="font-size:0.8rem;color:#9ca3af">' + inv.invoice_no + '</td>'
            + '<td><div style="font-weight:600">' + inv.customer_name + '</div>'
            + '<div style="font-size:0.75rem;color:#9ca3af">' + (inv.customer_phone || '') + '</div></td>'
            + '<td style="font-weight:600">' + formatCurrency(inv.grand_total) + '</td>'
            + '<td>' + formatCurrency(inv.amount_paid || 0) + '</td>'
            + '<td><span class="badge ' + statusClass + '">' + inv.payment_status + '</span></td>'
            + '<td><button class="btn btn-sm btn-outline" onclick="InvoicesPage.showForm(\'' + inv._id + '\')">✏️</button>'
            + '<button class="btn btn-sm btn-outline" onclick="InvoicesPage.printInvoice(\'' + inv._id + '\')">🖨️</button>'
            + '<button class="btn btn-sm btn-danger" onclick="InvoicesPage.deleteInvoice(\'' + inv._id + '\')">🗑️</button></td>'
            + '</tr>';
        }).join('');
      } else {
        html += '<tr><td colspan="6" class="text-center">No invoices found</td></tr>';
      }

      html += '</tbody></table></div></div>';

      if (data.pages > 1) {
        html += '<div class="flex justify-center gap-2" style="margin-top:1rem">'
          + '<button class="btn btn-sm btn-outline"' + (this.currentPage <= 1 ? ' disabled' : '') + ' onclick="InvoicesPage.loadPage(' + (this.currentPage - 1) + ')">Prev</button>'
          + '<span style="color:#9ca3af;padding:0.5rem">' + this.currentPage + ' / ' + data.pages + '</span>'
          + '<button class="btn btn-sm btn-outline"' + (this.currentPage >= data.pages ? ' disabled' : '') + ' onclick="InvoicesPage.loadPage(' + (this.currentPage + 1) + ')">Next</button>'
          + '</div>';
      }

      return html;
    } catch (err) {
      return '<div class="empty-state"><div class="icon">📄</div><h3>Failed to load invoices</h3></div>';
    }
  },

  loadPage(page) { this.currentPage = page; App.renderPage('invoices'); },
  searchInvoices(value) { this.search = value; this.currentPage = 1; clearTimeout(this._timer); this._timer = setTimeout(function() { App.renderPage('invoices'); }, 400); },
  filterStatus(value) { this.statusFilter = value; this.currentPage = 1; App.renderPage('invoices'); },

  async showForm(id) {
    if (!this._productsCache.length) {
      try { var res = await API.get('/products?limit=200'); this._productsCache = res.products; }
      catch (e) { this._productsCache = []; }
    }

    var items = [{ name: '', qty: 1, price: 0 }];
    var customerName = '', customerPhone = '', customerEmail = '', customerAddress = '';
    var discount = 0, tax = 0, amountPaid = 0, paymentMethod = 'Cash', dueDate = '', notes = '';

    if (id) {
      try {
        var res = await API.get('/invoices/' + id);
        var inv = res.invoice;
        customerName = inv.customer_name;
        customerPhone = inv.customer_phone;
        customerEmail = inv.customer_email;
        customerAddress = inv.customer_address || '';
        items = inv.items;
        discount = inv.discount || 0;
        tax = inv.tax || 0;
        amountPaid = inv.amount_paid || 0;
        paymentMethod = inv.payment_method || 'Cash';
        dueDate = inv.due_date || '';
        notes = inv.notes || '';
      } catch (e) {}
    }

    showModal(id ? 'Edit Invoice' : 'New Invoice',
      '<form id="invoice-form">'
      + '<input type="hidden" name="invoice_id" value="' + (id || '') + '">'
      + '<div class="form-row">'
      + '<div class="form-group"><label class="form-label">Customer Name *</label><input type="text" name="customer_name" value="' + customerName + '" required></div>'
      + '<div class="form-group"><label class="form-label">Phone</label><input type="tel" name="customer_phone" value="' + customerPhone + '"></div>'
      + '</div>'
      + '<div class="form-row">'
      + '<div class="form-group"><label class="form-label">Email</label><input type="email" name="customer_email" value="' + customerEmail + '"></div>'
      + '<div class="form-group"><label class="form-label">Due Date</label><input type="date" name="due_date" value="' + dueDate + '"></div>'
      + '</div>'
      + '<div class="form-group"><label class="form-label">Address</label><textarea name="customer_address">' + customerAddress + '</textarea></div>'
      + '<div class="form-group"><label class="form-label">Items</label>'
      + '<div id="inv-items-container"></div>'
      + '<button type="button" class="btn btn-sm btn-outline mt-2" onclick="InvoicesPage.addItem()">+ Add Item</button>'
      + '<button type="button" class="btn btn-sm btn-outline mt-2" onclick="InvoicesPage.addFromProducts()">📦 From Products</button>'
      + '</div>'
      + '<div class="form-row">'
      + '<div class="form-group"><label class="form-label">Discount (₹)</label><input type="number" step="0.01" name="discount" value="' + discount + '" oninput="InvoicesPage.calcTotal()"></div>'
      + '<div class="form-group"><label class="form-label">Tax (₹)</label><input type="number" step="0.01" name="tax" value="' + tax + '" oninput="InvoicesPage.calcTotal()"></div>'
      + '</div>'
      + '<div class="form-row">'
      + '<div class="form-group"><label class="form-label">Amount Paid (₹)</label><input type="number" step="0.01" name="amount_paid" value="' + amountPaid + '" oninput="InvoicesPage.calcTotal()"></div>'
      + '<div class="form-group"><label class="form-label">Payment Method</label><select name="payment_method">'
      + '<option value="Cash"' + (paymentMethod === 'Cash' ? ' selected' : '') + '>Cash</option>'
      + '<option value="UPI"' + (paymentMethod === 'UPI' ? ' selected' : '') + '>UPI</option>'
      + '<option value="Card"' + (paymentMethod === 'Card' ? ' selected' : '') + '>Card</option>'
      + '<option value="Bank Transfer"' + (paymentMethod === 'Bank Transfer' ? ' selected' : '') + '>Bank Transfer</option>'
      + '<option value="Cheque"' + (paymentMethod === 'Cheque' ? ' selected' : '') + '>Cheque</option>'
      + '</select></div></div>'
      + '<div style="text-align:right;padding:0.5rem 0;font-size:1.1rem;font-weight:700">'
      + 'Total: <span id="inv-total">₹0.00</span> | Balance: <span id="inv-balance">₹0.00</span></div>'
      + '<div class="form-group"><label class="form-label">Notes</label><textarea name="notes">' + notes + '</textarea></div>'
      + '<button type="submit" class="btn btn-primary btn-block">' + (id ? 'Update' : 'Create') + ' Invoice</button></form>'
    );

    var container = document.getElementById('inv-items-container');
    if (container) {
      items.forEach(function(item, i) { InvoicesPage.renderItemRow(container, item, i); });
      this.calcTotal();
    }

    document.getElementById('invoice-form')?.addEventListener('submit', async function(e) {
      e.preventDefault();
      var form = e.target;
      var data = {
        customer_name: form.customer_name.value,
        customer_phone: form.customer_phone.value,
        customer_email: form.customer_email.value,
        customer_address: form.customer_address.value,
        discount: parseFloat(form.discount.value) || 0,
        tax: parseFloat(form.tax.value) || 0,
        amount_paid: parseFloat(form.amount_paid.value) || 0,
        payment_method: form.payment_method.value,
        due_date: form.due_date.value,
        notes: form.notes.value,
        items: InvoicesPage.collectItems(),
      };
      if (!data.items.length) { showToast('Add at least one item', 'error'); return; }

      var iid = form.invoice_id.value;
      try {
        if (iid) { await API.put('/invoices/' + iid, data); showToast('Invoice updated'); }
        else { await API.post('/invoices', data); showToast('Invoice created'); }
        closeModal(); App.renderPage('invoices');
      } catch (err) { showToast(err.message, 'error'); }
    });
  },

  renderItemRow(container, item, i) {
    var prods = this._productsCache;
    var div = document.createElement('div');
    div.className = 'item-row';

    var opts = '<option value="">-- Select Product --</option>';
    prods.forEach(function(p) {
      var sel = p._id === (item.product_id || '') ? ' selected' : '';
      var label = p.name || p.sku || ('# ' + p._id);
      opts += '<option value="' + p._id + '" data-price="' + (p.selling_price || p.price) + '"' + sel + '>' + label + '</option>';
    });

    div.innerHTML = '<select onchange="InvoicesPage.onProductSelect(this)">' + opts + '</select>'
      + '<input type="number" class="item-qty" value="' + item.qty + '" min="1" oninput="InvoicesPage.calcTotal()">'
      + '<input type="number" step="0.01" class="item-price" value="' + item.price + '" oninput="InvoicesPage.calcTotal()">'
      + '<span class="item-total" style="font-weight:600;text-align:right">₹0</span>'
      + '<button type="button" class="btn btn-icon btn-danger btn-sm" onclick="this.closest(\'.item-row\').remove();InvoicesPage.calcTotal()">×</button>';

    container.appendChild(div);
    var qtyInput = div.querySelector('.item-qty');
    var priceInput = div.querySelector('.item-price');
    if (qtyInput) qtyInput.value = item.qty;
    if (priceInput) priceInput.value = item.price;
    this.calcTotal();
  },

  onProductSelect(select) {
    var opt = select.options[select.selectedIndex];
    var row = select.closest('.item-row');
    if (opt.value) { row.querySelector('.item-price').value = opt.dataset.price; }
    this.calcTotal();
  },

  addItem() {
    var container = document.getElementById('inv-items-container');
    if (container) this.renderItemRow(container, { name: '', qty: 1, price: 0 }, container.children.length);
  },

  addFromProducts() {
    var prods = this._productsCache;
    var container = document.getElementById('inv-items-container');
    if (!container) return;

    // Refresh cache from API if empty
    if (!prods || !prods.length) {
      showToast('No products found. Add products first.', 'error');
      return;
    }

    var listHtml = '<div style="max-height:50vh;overflow-y:auto">';
    prods.forEach(function(p) {
      var displayName = p.name || p.sku || ('Product #' + p._id);
      var val = JSON.stringify({ product_id: p._id, name: displayName, price: p.selling_price || p.price || 0 });
      listHtml += '<label style="display:flex;align-items:center;gap:0.5rem;padding:0.5rem;border-bottom:1px solid #2d2d4a;cursor:pointer">'
        + '<input type="checkbox" class="prod-select" value=\'' + val + '\'>'
        + '<div><div style="font-weight:600">' + displayName + '</div>'
        + '<div style="font-size:0.75rem;color:#9ca3af">' + (p.sku ? p.sku + ' | ' : '') + formatCurrency(p.selling_price || p.price || 0) + '</div></div></label>';
    });
    listHtml += '</div>'
      + '<div class="flex gap-2 mt-2">'
      + '<button type="button" class="btn btn-primary btn-block" onclick="InvoicesPage.addSelectedProducts()">Add Selected</button>'
      + '<button type="button" class="btn btn-danger" onclick="closeModal()">Cancel</button>'
      + '</div>';
    showModal('Select Products (' + prods.length + ' items)', listHtml, true);
  },

  addSelectedProducts() {
    var checks = document.querySelectorAll('.prod-select:checked');
    var container = document.getElementById('inv-items-container');
    if (!container) return;
    checks.forEach(function(c) {
      var p = JSON.parse(c.value);
      InvoicesPage.renderItemRow(container, { product_id: p.product_id, name: p.name, qty: 1, price: p.price }, container.children.length);
    });
    closeModal();
  },

  collectItems() {
    var rows = document.querySelectorAll('#inv-items-container .item-row');
    return Array.from(rows).map(function(row) {
      var select = row.querySelector('select');
      var name = select ? (select.options[select.selectedIndex] ? select.options[select.selectedIndex].text : '') : '';
      var qty = parseFloat(row.querySelector('.item-qty').value) || 1;
      var price = parseFloat(row.querySelector('.item-price').value) || 0;
      return { product_id: select ? select.value : '', name: name, qty: qty, price: price, total: qty * price };
    }).filter(function(i) { return i.name || i.price; });
  },

  calcTotal() {
    var items = this.collectItems();
    var subtotal = items.reduce(function(s, i) { return s + i.total; }, 0);
    var discount = parseFloat(document.querySelector('[name="discount"]')?.value) || 0;
    var tax = parseFloat(document.querySelector('[name="tax"]')?.value) || 0;
    var amountPaid = parseFloat(document.querySelector('[name="amount_paid"]')?.value) || 0;
    var total = subtotal - discount + tax;
    var balance = total - amountPaid;

    var totals = document.querySelectorAll('#inv-items-container .item-total');
    items.forEach(function(item, idx) {
      if (totals[idx]) totals[idx].textContent = formatCurrency(item.total);
    });

    var totalEl = document.getElementById('inv-total');
    var balanceEl = document.getElementById('inv-balance');
    if (totalEl) totalEl.textContent = formatCurrency(total);
    if (balanceEl) balanceEl.textContent = formatCurrency(balance);
  },

  async deleteInvoice(id) {
    if (!confirm('Delete this invoice?')) return;
    try { await API.delete('/invoices/' + id); showToast('Invoice deleted'); App.renderPage('invoices'); }
    catch (err) { showToast(err.message, 'error'); }
  },

  async printInvoice(id) {
    try {
      var blob = await API.download('/invoices/' + id + '/pdf');
      downloadBlob(blob, 'Invoice-' + id + '.pdf');
    } catch (err) { showToast(err.message, 'error'); }
  },

  async exportExcel() {
    try { var blob = await API.download('/invoices/export'); downloadBlob(blob, 'invoices.xlsx'); showToast('Exported successfully'); }
    catch (err) { showToast(err.message, 'error'); }
  },

  init() {},
};

File: backend/routes/invoices.py
from datetime import datetime, date
from flask import Blueprint, request, jsonify, current_app, send_file
from flask_jwt_extended import jwt_required, get_jwt
from utils.helpers import admin_required, employee_or_admin, serialize_doc, serialize_list
from utils.pdf_utils import generate_document_pdf, format_currency
from bson.objectid import ObjectId
import io
import openpyxl


def generate_invoice_no(db):
    today = date.today().strftime('%Y%m%d')
    last = db.invoices.find_one(
        {'invoice_no': {'$regex': f'^INV-{today}'}},
        sort=[('invoice_no', -1)]
    )
    if last:
        last_num = int(last['invoice_no'].split('-')[-1])
        new_num = last_num + 1
    else:
        new_num = 1
    return f'INV-{today}-{new_num:04d}'


invoices_bp = Blueprint('invoices', __name__)


@invoices_bp.route('', methods=['GET'])
@jwt_required()
def get_invoices():
    current_user = get_jwt()
    db = current_app.config['db']

    page = int(request.args.get('page', 1))
    limit = int(request.args.get('limit', 50))
    search = request.args.get('search', '')
    status = request.args.get('status', '')

    query = {}
    if current_user['role'] == 'employee':
        query['created_by'] = current_user['full_name']
    if search:
        query['$or'] = [
            {'invoice_no': {'$regex': search, '$options': 'i'}},
            {'customer_name': {'$regex': search, '$options': 'i'}},
        ]
    if status:
        query['payment_status'] = status

    total = db.invoices.count_documents(query)
    invoices = list(db.invoices.find(query)
                    .sort('created_at', -1)
                    .skip((page - 1) * limit)
                    .limit(limit))

    return jsonify({
        'invoices': serialize_list(invoices),
        'total': total,
        'page': page,
        'pages': (total + limit - 1) // limit,
    })


@invoices_bp.route('', methods=['POST'])
@employee_or_admin
def create_invoice():
    current_user = get_jwt()
    data = request.get_json()
    db = current_app.config['db']

    items = data.get('items', [])
    subtotal = sum(float(item.get('qty', 1)) * float(item.get('price', 0)) for item in items)
    discount = float(data.get('discount', 0))
    tax = float(data.get('tax', 0))
    grand_total = subtotal - discount + tax
    amount_paid = float(data.get('amount_paid', 0))
    balance = grand_total - amount_paid

    invoice = {
        'invoice_no': generate_invoice_no(db),
        'customer_name': data.get('customer_name', ''),
        'customer_phone': data.get('customer_phone', ''),
        'customer_email': data.get('customer_email', ''),
        'customer_address': data.get('customer_address', ''),
        'items': items,
        'subtotal': subtotal,
        'discount': discount,
        'tax': tax,
        'grand_total': grand_total,
        'amount_paid': amount_paid,
        'balance': balance,
        'payment_status': 'Paid' if balance <= 0 else ('Partial' if amount_paid > 0 else 'Unpaid'),
        'payment_method': data.get('payment_method', 'Cash'),
        'due_date': data.get('due_date', ''),
        'notes': data.get('notes', ''),
        'status': 'Active',
        'created_by': current_user['full_name'],
        'created_at': datetime.utcnow(),
        'updated_at': datetime.utcnow(),
    }

    result = db.invoices.insert_one(invoice)
    invoice['_id'] = str(result.inserted_id)

    return jsonify({'message': 'Invoice created', 'invoice': invoice}), 201


@invoices_bp.route('/<invoice_id>', methods=['GET'])
@jwt_required()
def get_invoice(invoice_id):
    db = current_app.config['db']
    invoice = db.invoices.find_one({'_id': ObjectId(invoice_id)})
    if not invoice:
        return jsonify({'error': 'Invoice not found'}), 404
    return jsonify({'invoice': serialize_doc(invoice)})


@invoices_bp.route('/<invoice_id>', methods=['PUT'])
@jwt_required()
def update_invoice(invoice_id):
    current_user = get_jwt()
    data = request.get_json()
    db = current_app.config['db']

    invoice = db.invoices.find_one({'_id': ObjectId(invoice_id)})
    if not invoice:
        return jsonify({'error': 'Invoice not found'}), 404

    if current_user['role'] == 'employee' and invoice.get('created_by') != current_user['full_name']:
        return jsonify({'error': 'Not authorized'}), 403

    update = {}
    for field in ['customer_name', 'customer_phone', 'customer_email', 'customer_address',
                  'payment_method', 'due_date', 'notes', 'status']:
        if field in data:
            update[field] = data[field]

    if 'items' in data:
        items = data['items']
        subtotal = sum(float(item.get('qty', 1)) * float(item.get('price', 0)) for item in items)
        discount = float(data.get('discount', invoice.get('discount', 0)))
        tax = float(data.get('tax', invoice.get('tax', 0)))
        grand_total = subtotal - discount + tax
        amount_paid = float(data.get('amount_paid', invoice.get('amount_paid', 0)))
        balance = grand_total - amount_paid
        update['items'] = items
        update['subtotal'] = subtotal
        update['discount'] = discount
        update['tax'] = tax
        update['grand_total'] = grand_total
        update['amount_paid'] = amount_paid
        update['balance'] = balance
        update['payment_status'] = 'Paid' if balance <= 0 else ('Partial' if amount_paid > 0 else 'Unpaid')
    elif 'amount_paid' in data:
        amount_paid = float(data['amount_paid'])
        grand_total = invoice.get('grand_total', 0)
        balance = grand_total - amount_paid
        update['amount_paid'] = amount_paid
        update['balance'] = balance
        update['payment_status'] = 'Paid' if balance <= 0 else ('Partial' if amount_paid > 0 else 'Unpaid')

    update['updated_at'] = datetime.utcnow()
    db.invoices.update_one({'_id': ObjectId(invoice_id)}, {'$set': update})
    invoice = db.invoices.find_one({'_id': ObjectId(invoice_id)})

    return jsonify({'message': 'Invoice updated', 'invoice': serialize_doc(invoice)})


@invoices_bp.route('/<invoice_id>', methods=['DELETE'])
@admin_required
def delete_invoice(invoice_id):
    db = current_app.config['db']
    db.invoices.delete_one({'_id': ObjectId(invoice_id)})
    return jsonify({'message': 'Invoice deleted'})


@invoices_bp.route('/<invoice_id>/pdf', methods=['GET'])
@jwt_required()
def invoice_pdf(invoice_id):
    db = current_app.config['db']
    inv = db.invoices.find_one({'_id': ObjectId(invoice_id)})
    if not inv:
        return jsonify({'error': 'Invoice not found'}), 404

    customer_lines = [inv.get('customer_name', ''), inv.get('customer_phone', ''), inv.get('customer_email', '')]
    if inv.get('customer_address'):
        customer_lines.append(inv['customer_address'])

    extra_lines = ['Status: ' + (inv.get('payment_status') or '')]
    if inv.get('due_date'):
        extra_lines.append('Due: ' + str(inv['due_date']))
    extra_lines += [
        'Payment: ' + (inv.get('payment_method') or ''),
        'Paid: ' + format_currency(inv.get('amount_paid', 0)),
        'Balance: ' + format_currency(inv.get('balance', 0)),
    ]

    totals = {
        'subtotal': inv.get('subtotal', 0),
        'discount': inv.get('discount', 0),
        'tax': inv.get('tax', 0),
        'grand_total': inv.get('grand_total', 0),
    }

    buf = generate_document_pdf(
        'INVOICE',
        inv.get('invoice_no', ''),
        str(inv.get('created_at', '') or '')[:10],
        customer_lines,
        inv.get('items', []),
        totals,
        extra_lines,
    )

    return send_file(
        buf,
        mimetype='application/pdf',
        as_attachment=True,
        download_name=(inv.get('invoice_no', 'invoice') + '.pdf'),
    )


@invoices_bp.route('/export', methods=['GET'])
@admin_required
def export_invoices():
    db = current_app.config['db']
    invoices = list(db.invoices.find().sort('created_at', -1))

    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = 'Invoices'
    ws.append(['Invoice No', 'Customer', 'Phone', 'Amount', 'Paid', 'Balance', 'Status', 'Payment', 'Date'])

    for inv in invoices:
        ws.append([
            inv.get('invoice_no', ''),
            inv.get('customer_name', ''),
            inv.get('customer_phone', ''),
            inv.get('grand_total', 0),
            inv.get('amount_paid', 0),
            inv.get('balance', 0),
            inv.get('payment_status', ''),
            inv.get('payment_method', ''),
            str(inv.get('created_at', '') or ''),
        ])

    output = io.BytesIO()
    wb.save(output)
    output.seek(0)

    return send_file(
        output,
        mimetype='application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        as_attachment=True,
        download_name='invoices.xlsx'
    )
